x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStopDatafeedAction.java

```diff
@@ -194,21 +195,36 @@ private void normalStopDatafeed(Task task, StopDatafeedAction.Request request, A
                     r -> auditDatafeedStopped(datafeedTask),
                     e -> logger.error("[" + datafeedId + "] failed to remove task to stop unassigned datafeed", e))
                 );
+                allDataFeedsToWaitFor.add(datafeedTask);
             }
         }
 
+        for (String datafeedId : stoppingDatafeeds) {
+            PersistentTasksCustomMetadata.PersistentTask<?> datafeedTask = MlTasks.getDatafeedTask(datafeedId, tasks);
+            assert datafeedTask != null : "Requested datafeed [" + datafeedId + "] be stopped, but datafeed's task could not be found.";
+            allDataFeedsToWaitFor.add(datafeedTask);
+        }
+
         request.setNodes(executorNodes.toArray(new String[0]));
 
-        // wait for started and stopping datafeeds
-        // Map datafeedId -> datafeed task Id.
-        List<String> allDataFeedsToWaitFor = Stream.concat(
-                startedDatafeeds.stream().map(MlTasks::datafeedTaskId),
-                stoppingDatafeeds.stream().map(MlTasks::datafeedTaskId))
-                .collect(Collectors.toList());
+        final Set<String> movedDatafeeds = new HashSet<>();
```

Comment:
Does this need to be a `ConcurrentHashSet` or whatever because the collection is periodically updated/read via another thread?

Reply:
I _think_ it should be OK as-is, because all the insertions are done in the cluster state listener thread, and then that will also call the listener where the contents are used.  However, since it will be very rare that anything gets put in this map it doesn't really hurt to make it a concurrent set just in case the usage pattern changes in the future.